src/components/shared/StateProvider.tsx

```diff
@@ -0,0 +1,59 @@
+import { createContext, Dispatch, PropsWithChildren, useContext, useReducer } from 'react';
+
+interface ApplicationState {
+    expandedSykmeldte: string[];
+}
+
+const defaultState: ApplicationState = {
+    expandedSykmeldte: [],
+};
+
+type ToggleExpandSykmeldte = {
+    type: 'toggleExpandSykmeldte';
```

Comment:
Kan bruke standard uppercase på actionTypes? `TOGGLE_EXPAND_SYKMELDTE` 
Og kanskje ha det som ein const? `const TOGGLE_EXPAND_SYKMELDTE = 'TOGGLE_EXPAND_SYKMELDTE';`

Reply:
Det er ikke en string, det er en string literal (https://www.typescriptlang.org/docs/handbook/literal-types.html). Det er så unionen på linje 20 får en enum-aktig `type` som kan brukes til å diskriminere mellom to typer.

Det blir på en måte som at typen er:

```
type CombinedType = {
  type: 'clearAll' | 'toggleExpandSykmeldte'
  payload?: string
}
```

Bare at TypeScript er i stand til å se at dersom type er `clearAll` så er payload ikke tilstede, og dersom type er `toggleExpandSykmeldte` så er payload non-nullable `string`. Det er en ganske kraftig feature i TS å kunne diskriminere sånn. :smile:

Reply:
Skjønnar! Men det vil fortsatt vere string literal med mitt forslag? Eller misforstår eg? 

```
const TOGGLE_EXPAND_SYKMELDTE = 'TOGGLE_EXPAND_SYKMELDTE';

type ToggleExpandSykmeldte = {
    type: TOGGLE_EXPAND_SYKMELDTE;
    payload: string;
};
```

Og `type` vil vere `TOGGLE_EXPAND_SYKMELDTE` og ikkje string?

Reply:
Man kan ikke referere til en verdi i en type sånn, man kan bruke `typeof` sånn:

![image](https://user-images.githubusercontent.com/1507032/148227462-42c8d080-3386-4116-a5cb-a454e7799f77.png)

Da ser du den inferrer samme string literal. Men det gir oss ikke noe i form av å hindre typos/sterkere typing, fordi unions av string literals fungerer som mini-enums som passer på at man gjør det rett.

![image](https://user-images.githubusercontent.com/1507032/148227786-1999a749-6b07-4f7f-8499-708506f5fbde.png)

Så da sparer man seg en del sånn boilerplate som man vanligvis ville gjort i JavaScript, fordi magic strings i et utypet språk er jo dødsfarlig. :smile:

Reply:
Okey, skjønner! Takk for forklaring!